compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -815,16 +816,13 @@ class SpaceEngine(using Context) extends SpaceLogic {
         isCheckable(and.tp1) || isCheckable(and.tp2)
       }) ||
       tpw.isRef(defn.BooleanClass) ||
-      classSym.isAllOf(JavaEnumTrait)
+      classSym.isAllOf(JavaEnumTrait) ||
+      classSym.is(Case) && productSelectorTypes(tpw, sel.srcPos).exists(isCheckable(_))
+    })
```

Comment:
I tried to understand why the cache breaks the cycle but couldn't figure it out --- when the recursive call tries to use the cache, the value is not there yet, so it will run into cycles.

But obviously, it works. Could you give a hint? :smile:

Reply:
I had exactly the same thought after I sent the PR and didn't understand it either.

But I think I do now: because the test case for #12488 is running under `-Ycheck-all-patmat` so it's short-circuting...  Damn `-Ycheck-all-patmat`.  Why does that flag exist?  It's misleading test results, as I see it.

Reply:
It's created a long time ago for testing whether the algorithm works generally. Now it becomes a "feature".

Reply:
I only partially understood your answer there.  Do you think it should continue to exist?  In Scala 2 the only equivalents we have are: `-Xlint:strict-unsealed-patmat` to consider unsealed scrutinees checkable and `-Xnon-strict-patmat-analysis` to disable the pessimistic approximation of guards and custom extractors not matching (the "strict" patmat analysis).

Perhaps looking to transition the tests and the logic out from under the `-Ycheck-all-patmat` would be a good project to get me pinging around the patmat code, learning parts of it.

Reply:
> Do you think it should continue to exist? 

I don't really have an opinion here. It's not supposed to be used by end-users. So there is room to change.